### ISSUE TITLE
UnusedVariable false positive

### ISSUE THREAD

**Issue by iluuu1994** [vimeo/psalm#2052]
Please ignore the horrific code:

![image](https://user-images.githubusercontent.com/1752683/63591484-3b848500-c5af-11e9-8e45-357763813f05.png)

> Variable $lang is never referenced

I get a lot of these false positives in our legacy code base for both key and value in `foreach` statements. Unfortunately I was not yet able to reproduce this on https://psalm.dev.

Any clues?

**Comment by iluuu1994** [comment]
Not exclusive to `foreach` statements:

![image](https://user-images.githubusercontent.com/1752683/63591737-e8f79880-c5af-11e9-9c87-c4b086727d9f.png)

> Variable $translationCode is never referenced

**Comment by iluuu1994** [comment]
Or here:

![image](https://user-images.githubusercontent.com/1752683/63592084-011be780-c5b1-11e9-999f-197e34e2a7df.png)

> Variable $qAlias is never referenced

**Comment by muglug** [comment]
@iluuu1994 the last is reproducible with

```php
/**
 * @psalm-suppress MixedMethodCall
 */
function foo(string $s, object $o) : void {
    $o->foo("COUNT{$s}");
}
```

**Comment by iluuu1994** [comment]
Cool! Hopefully all the other cases have the same origin.

**Comment by muglug** [comment]
To diagnose I'd recommend turning on `totallyTyped="true"` - I reckon at least some of the other issues are to do with `mixed` types being inferred, and Psalm handling those improperly.

**Comment by iluuu1994** [comment]
![image](https://user-images.githubusercontent.com/1752683/63603109-359b9d80-c5c9-11e9-95f4-a34798e77969.png)

Ok, I'll report back in a few months 😆

**Comment by muglug** [comment]
Haha - mainly interested what types were mixed when the unused variable issue was reported

**Comment by iluuu1994** [comment]
Managed to reproduce one more case:

https://psalm.dev/r/837ac3b930

**Comment by iluuu1994** [comment]
A lot of our cases could have something to do with reserved keywords. There's a class called `Resource` in our codebase. This works fine in PHP but is definitely a dumb idea.

This case for example only occurs when the class is called `Resource`. Renaming it to `Resource_` fixes the problem.

`Resource`: https://psalm.dev/r/8bc31a6624
`Resource_`: https://psalm.dev/r/23073dc0d3

**Comment by iluuu1994** [comment]
Here's another one (unrelated to reserved keywords):

https://psalm.dev/r/c8e4b5ab2b

**Comment by iluuu1994** [comment]
Another one:

https://psalm.dev/r/789a184426

**Comment by iluuu1994** [comment]
That should be all of them. All others were legit or variations of the ones already reproduced. I made a list at the top so we can track them all.

**Comment by muglug** [comment]
Thanks! I think the resource issue can just be fixed by https://github.com/vimeo/psalm/compare/master...muglug-patch-1 (though I’m on mobile, won’t be able to verify until evening)

**Comment by iluuu1994** [comment]
@muglug Please don't worry, enjoy your day off! I'm just documenting my findings 🙂

**Comment by iluuu1994** [comment]
@muglug The `ReservedWord` one is still broken.

**Comment by muglug** [comment]
Add a namespace!

My change was to allow it except in the root namespace, which I think is fair?

**Comment by iluuu1994** [comment]
@muglug Btw, I'm not complaining about the `ReservedWord` 🙂 It's fine, our codebase is stupid. I'll gladly ignore that error. But that error causes the unrelated `UnusedVariable` errors that I can't ignore as easily.

**Comment by iluuu1994** [comment]
Thanks @muglug!

**Comment by iluuu1994** [comment]
Great! Thank you!
